### ISSUE TITLE
Various JID parsing errors

### ISSUE THREAD

**Issue by SamWhited** [jackal-xmpp/stravaganza#2]
[RFC 7622] specified exactly how JIDs should be parsed, but this package does not follow those rules, resulting in panics and JIDs being incorrectly labeled as valid or invalid.

For example, parsing the JID `example.net/foobar@test` results in the following:

```
panic: runtime error: slice bounds out of range [19:11] [recovered]
        panic: runtime error: slice bounds out of range [19:11]
```

And parsing `/foobar@test` results in a valid JID (even though there is no domainpart because you must parse the resourcepart first which makes this whole JID a resourcepart in this case).

Here is a test (borrowed mostly from `mellium.im/xmpp/jid`, but I give you permission to use it however you want with no restrictions, etc.) with some invalid JIDs (some of which pass, some of which don't) that may be a bit more comprehensive than what's there now if you'd like to use it when updating:

```go
var invalidutf8 = string([]byte{0xff, 0xfe, 0xfd})

var invalidJIDs = [...]string{
	0:  "test@/test",
	1:  invalidutf8 + "@example.com/rp",
	2:  invalidutf8 + "/rp",
	3:  invalidutf8,
	4:  "example.com/" + invalidutf8,
	5:  "lp@/rp",
	6:  `b"d@example.net`,
	7:  `b&d@example.net`,
	8:  `b'd@example.net`,
	9:  `b:d@example.net`,
	10: `b<d@example.net`,
	11: `b>d@example.net`,
	12: `e@example.net/`,
	13: `@example.net/`,
	14: `foo bar@example.com`,
	15: `henryⅣ@example.com`,
	16: `♚@example.com`,
	17: `juliet@`,
	18: `"juliet"@example.com`,
	19: `/foobar`,
	20: `/foobar@test`,
	21: `example.net/foobar@test`,
}

func TestInvalidParseJIDs(t *testing.T) {
	for i, tc := range invalidJIDs {
		t.Run(strconv.Itoa(i), func(t *testing.T) {
			_, err := jid.NewWithString(tc, false)
			if err == nil {
				t.Errorf("expected JID %s to fail", tc)
			}
		})
	}
}
```

[RFC 7622]: https://tools.ietf.org/html/rfc7622

**Comment by ortuman** [comment]
Thanks again Sam for the invaluable advice.

I'll borrow this test and see to integrate it into the project's test battery as soon as possible.

I'll let you know here when done. 🙇

**Comment by ortuman** [comment]
👋 hello again @SamWhited 

I see there's also a [splitString](https://github.com/mellium/xmpp/blob/bdb9a80731af29ffb23dbebdf45dc4e757d6716d/jid/jid.go#L340-L403) implementation in the `mellium.im/xmpp/jid` repository that applies most parsing validations, and I was wondering if I could also borrow this piece of code (naming the original repository ofc) to import it.  WDYT?

**Comment by SamWhited** [comment]
Sure, feel free to re-license that function under Apache 2.

**Comment by ortuman** [comment]
Fixed @ e19488f2704e80d30890f7a1285ddeff769e8e01

Thanks! 🙌
